File: C++ General/CANdle-MultiAnimation/src/main/cpp/subsystems/CANdleSystem.cpp
/**
 * Phoenix Software License Agreement
 *
 * Copyright (C) Cross The Road Electronics.  All rights
 * reserved.
 * 
 * Cross The Road Electronics (CTRE) licenses to you the right to 
 * use, publish, and distribute copies of CRF (Cross The Road) firmware files (*.crf) and 
 * Phoenix Software API Libraries ONLY when in use with CTR Electronics hardware products
 * as well as the FRC roboRIO when in use in FRC Competition.
 * 
 * THE SOFTWARE AND DOCUMENTATION ARE PROVIDED "AS IS" WITHOUT
 * WARRANTY OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT
 * LIMITATION, ANY WARRANTY OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE, TITLE AND NON-INFRINGEMENT. IN NO EVENT SHALL
 * CROSS THE ROAD ELECTRONICS BE LIABLE FOR ANY INCIDENTAL, SPECIAL, 
 * INDIRECT OR CONSEQUENTIAL DAMAGES, LOST PROFITS OR LOST DATA, COST OF
 * PROCUREMENT OF SUBSTITUTE GOODS, TECHNOLOGY OR SERVICES, ANY CLAIMS
 * BY THIRD PARTIES (INCLUDING BUT NOT LIMITED TO ANY DEFENSE
 * THEREOF), ANY CLAIMS FOR INDEMNITY OR CONTRIBUTION, OR OTHER
 * SIMILAR COSTS, WHETHER ASSERTED ON THE BASIS OF CONTRACT, TORT
 * (INCLUDING NEGLIGENCE), BREACH OF WARRANTY, OR OTHERWISE
 */
#include <iostream>
#include "subsystems/CANdleSystem.h"
#include "ctre/phoenix/led/ColorFlowAnimation.h"
#include "ctre/phoenix/led/FireAnimation.h"
#include "ctre/phoenix/led/LarsonAnimation.h"
#include "ctre/phoenix/led/RainbowAnimation.h"
#include "ctre/phoenix/led/RgbFadeAnimation.h"
#include "ctre/phoenix/led/SingleFadeAnimation.h"
#include "ctre/phoenix/led/StrobeAnimation.h"
#include "ctre/phoenix/led/TwinkleAnimation.h"
#include "ctre/phoenix/led/TwinkleOffAnimation.h"

using namespace ctre::phoenix::led;

CANdleSystem::CANdleSystem() {
    ChangeAnimation(AnimationTypes::SetAll);
    CANdleConfiguration configAll {};
    configAll.statusLedOffWhenActive = true;
    configAll.disableWhenLOS = false;
    configAll.stripType = LEDStripType::GRB;
    configAll.brightnessScalar = 0.1;
    configAll.vBatOutputMode = VBatOutputMode::Modulated;
    m_candle.ConfigAllSettings(configAll, 100);
}

CANdleSystem::~CANdleSystem() {
    if(m_toAnimate != NULL) delete m_toAnimate;
}

void CANdleSystem::IncrementAnimation() {
    switch(m_currentAnimation) {
        case ColorFlow: ChangeAnimation(AnimationTypes::Fire); break;
        case Fire: ChangeAnimation(AnimationTypes::Larson); break;
        case Larson: ChangeAnimation(AnimationTypes::Rainbow); break;
        case Rainbow: ChangeAnimation(AnimationTypes::RgbFade); break;
        case RgbFade: ChangeAnimation(AnimationTypes::SingleFade); break;
        case SingleFade: ChangeAnimation(AnimationTypes::Strobe); break;
        case Strobe: ChangeAnimation(AnimationTypes::Twinkle); break;
        case Twinkle: ChangeAnimation(AnimationTypes::TwinkleOff); break;
        case TwinkleOff: ChangeAnimation(AnimationTypes::ColorFlow); break;
        case SetAll: ChangeAnimation(AnimationTypes::ColorFlow); break;
    }
}
void CANdleSystem::DecrementAnimation() {
    switch(m_currentAnimation) {
        case ColorFlow: ChangeAnimation(AnimationTypes::TwinkleOff); break;
        case Fire: ChangeAnimation(AnimationTypes::ColorFlow); break;
        case Larson: ChangeAnimation(AnimationTypes::Fire); break;
        case Rainbow: ChangeAnimation(AnimationTypes::Larson); break;
        case RgbFade: ChangeAnimation(AnimationTypes::Rainbow); break;
        case SingleFade: ChangeAnimation(AnimationTypes::RgbFade); break;
        case Strobe: ChangeAnimation(AnimationTypes::SingleFade); break;
        case Twinkle: ChangeAnimation(AnimationTypes::Strobe); break;
        case TwinkleOff: ChangeAnimation(AnimationTypes::Twinkle); break;
        case SetAll: ChangeAnimation(AnimationTypes::ColorFlow); break;
    }
}
void CANdleSystem::SetColors() {
    ChangeAnimation(AnimationTypes::SetAll);
}
void CANdleSystem::Toggle5VOverride() {
    m_candle.configV5Enabled(m_last5V);
    m_last5V = !m_last5V;
}
void CANdleSystem::ToggleAnimDirection() {
    m_animDirection = !m_animDirection;
}
int CANdleSystem::GetMaximumAnimationCount() {
    return m_candle.GetMaxSimultaneousAnimationCount();
}


void CANdleSystem::UpdateSetLed(std::function<double()> r, std::function<double()> g, std::function<double()> b, std::function<double()> modulatedVbat) {
    m_r = r() * 255;
    m_g = g() * 255;
    m_b = b() * 255;
    m_modulatedOut = modulatedVbat();
}

/* Wrappers so we can access the CANdle from the subsystem */
double CANdleSystem::GetVbat() { return m_candle.GetBusVoltage(); }
double CANdleSystem::Get5V() { return m_candle.Get5VRailVoltage(); }
double CANdleSystem::GetCurrent() { return m_candle.GetCurrent(); }
double CANdleSystem::GetTemperature() { return m_candle.GetTemperature(); }
double CANdleSystem::GetModulatedOutput() { return m_candle.GetVBatModulation(); }
void CANdleSystem::ConfigBrightness(double percent) { m_candle.ConfigBrightnessScalar(percent, 0); }
void CANdleSystem::ConfigLos(bool disableWhenLos) { m_candle.ConfigLOSBehavior(disableWhenLos, 0); }
void CANdleSystem::ConfigLedType(LEDStripType type) { m_candle.ConfigLEDType(type, 0); }
void CANdleSystem::ConfigStatusLedBehavior(bool offWhenActive) { m_candle.ConfigStatusLedState(offWhenActive, 0); }

void CANdleSystem::ChangeAnimation(AnimationTypes toChange) {
    m_currentAnimation = toChange;
    
    if(m_toAnimate != NULL) delete m_toAnimate;

    switch(toChange)
    {
            default:
            case ColorFlow:
                m_candleChannel = 0;
                m_toAnimate = new ColorFlowAnimation(128, 20, 70, 0, 0.7, LEDS_PER_ANIMATION, ColorFlowAnimation::Direction::Forward, m_candleChannel * LEDS_PER_ANIMATION + 8);
                break;
            case Fire:
                m_candleChannel = 1;
                m_toAnimate = new FireAnimation(0.5, 0.7, LEDS_PER_ANIMATION, 0.8, 0.5, m_animDirection, m_candleChannel * LEDS_PER_ANIMATION + 8);
                break;
            case Larson:
                m_candleChannel = 2;
                m_toAnimate = new LarsonAnimation(0, 255, 46, 0, 0.1, LEDS_PER_ANIMATION, LarsonAnimation::BounceMode::Front, 3, m_candleChannel * LEDS_PER_ANIMATION + 8);
                break;
            case Rainbow:
                m_candleChannel = 3;
                m_toAnimate = new RainbowAnimation(1, 0.7, LEDS_PER_ANIMATION, m_animDirection, m_candleChannel * LEDS_PER_ANIMATION + 8);
                break;
            case RgbFade:
                m_candleChannel = 4;
                m_toAnimate = new RgbFadeAnimation(0.7, 0.4, LEDS_PER_ANIMATION, m_candleChannel * LEDS_PER_ANIMATION + 8);
                break;
            case SingleFade:
                m_candleChannel = 5;
                m_toAnimate = new SingleFadeAnimation(50, 2, 200, 0, 0.5, LEDS_PER_ANIMATION, m_candleChannel * LEDS_PER_ANIMATION + 8);
                break;
            case Strobe:
                m_candleChannel = 6;
                m_toAnimate = new StrobeAnimation(240, 10, 180, 0, 0.01, LEDS_PER_ANIMATION, m_candleChannel * LEDS_PER_ANIMATION + 8);
                break;
            case Twinkle:
                m_candleChannel = 7;
                m_toAnimate = new TwinkleAnimation(30, 70, 60, 0, 0.4, LEDS_PER_ANIMATION, TwinkleAnimation::TwinklePercent::Percent42, m_candleChannel * LEDS_PER_ANIMATION + 8);
                break;
            case TwinkleOff:
                m_candleChannel = 8;
                m_toAnimate = new TwinkleOffAnimation(70, 90, 175, 0, 0.2, LEDS_PER_ANIMATION, TwinkleOffAnimation::TwinkleOffPercent::Percent76, m_candleChannel * LEDS_PER_ANIMATION + 8);
                break;

            case SetAll:
                m_toAnimate = NULL;
                break;
    }
    std::cout << "Changed to " << std::to_string(m_currentAnimation) << std::endl;
}
void CANdleSystem::ClearAllAnims() {m_clearAllAnims = true;}

void CANdleSystem::Periodic() {
    // This method will be called once per scheduler run
    if(m_toAnimate == NULL) {
        if(!m_setAnim) {
            /* Only setLEDs once, because every set will transmit a frame */
            m_candle.SetLEDs(255, 255, 255, 0, 0, 1);
            m_candle.SetLEDs(255, 255, 0, 0, 1, 1);
            m_candle.SetLEDs(255, 0, 255, 0, 2, 1);
            m_candle.SetLEDs(255, 0, 0, 0, 3, 1);
            m_candle.SetLEDs(0, 255, 255, 0, 4, 1);
            m_candle.SetLEDs(0, 255, 0, 0, 5, 1);
            m_candle.SetLEDs(0, 0, 0, 0, 6, 1);
            m_candle.SetLEDs(0, 0, 255, 0, 7, 1);
            m_setAnim = true;
        }
    } else {
        m_candle.Animate(*m_toAnimate, m_candleChannel);
        m_setAnim = false;
    }

    if(m_clearAllAnims) {
        m_clearAllAnims = false;
        for(int i = 0; i < 10; ++i) {
            m_candle.ClearAnimation(i); // If using the 5.21.2 build of Phoenix, comment this out as there is a bug that causes a compilation error
                                        // This bug has been fixed in dev build 5.21.2-12-gfdb3388 and any full release after 5.21.2
        }
    }
}

void CANdleSystem::SimulationPeriodic() {
    // This method will be called once per scheduler run during simulation
}

File: Java General/CANdle MultiAnimation/src/main/java/frc/robot/subsystems/CANdleSystem.java
/**
 * Phoenix Software License Agreement
 *
 * Copyright (C) Cross The Road Electronics.  All rights
 * reserved.
 * 
 * Cross The Road Electronics (CTRE) licenses to you the right to 
 * use, publish, and distribute copies of CRF (Cross The Road) firmware files (*.crf) and 
 * Phoenix Software API Libraries ONLY when in use with CTR Electronics hardware products
 * as well as the FRC roboRIO when in use in FRC Competition.
 * 
 * THE SOFTWARE AND DOCUMENTATION ARE PROVIDED "AS IS" WITHOUT
 * WARRANTY OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT
 * LIMITATION, ANY WARRANTY OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE, TITLE AND NON-INFRINGEMENT. IN NO EVENT SHALL
 * CROSS THE ROAD ELECTRONICS BE LIABLE FOR ANY INCIDENTAL, SPECIAL, 
 * INDIRECT OR CONSEQUENTIAL DAMAGES, LOST PROFITS OR LOST DATA, COST OF
 * PROCUREMENT OF SUBSTITUTE GOODS, TECHNOLOGY OR SERVICES, ANY CLAIMS
 * BY THIRD PARTIES (INCLUDING BUT NOT LIMITED TO ANY DEFENSE
 * THEREOF), ANY CLAIMS FOR INDEMNITY OR CONTRIBUTION, OR OTHER
 * SIMILAR COSTS, WHETHER ASSERTED ON THE BASIS OF CONTRACT, TORT
 * (INCLUDING NEGLIGENCE), BREACH OF WARRANTY, OR OTHERWISE
 */

/**
 * Description:
 * The CANdle MultiAnimation example demonstrates using multiple animations with CANdle.
 * This example has the robot using a Command Based template to control the CANdle.
 * 
 * This example uses:
 * - A CANdle wired on the CAN Bus, with a 5m led strip attached for the extra animatinos.
 * 
 * Controls (with Xbox controller):
 * Right Bumper: Increment animation
 * Left Bumper: Decrement animation
 * Start Button: Switch to setting the first 8 LEDs a unique combination of colors
 * POV Right: Configure maximum brightness for the CANdle
 * POV Down: Configure medium brightness for the CANdle
 * POV Left: Configure brightness to 0 for the CANdle
 * POV Up: Change the direction of Rainbow and Fire, must re-select the animation to take affect
 * A: Print the VBat voltage in Volts
 * B: Print the 5V voltage in Volts
 * X: Print the current in amps
 * Y: Print the temperature in degrees C
 * 
 * Supported Version:
 * 	- CANdle: 22.1.1.0
 */

package frc.robot.subsystems;

import edu.wpi.first.wpilibj.XboxController;
import edu.wpi.first.wpilibj2.command.SubsystemBase;
import frc.robot.Constants;

import com.ctre.phoenix.led.*;
import com.ctre.phoenix.led.CANdle.LEDStripType;
import com.ctre.phoenix.led.CANdle.VBatOutputMode;
import com.ctre.phoenix.led.ColorFlowAnimation.Direction;
import com.ctre.phoenix.led.LarsonAnimation.BounceMode;
import com.ctre.phoenix.led.TwinkleAnimation.TwinklePercent;
import com.ctre.phoenix.led.TwinkleOffAnimation.TwinkleOffPercent;

public class CANdleSystem extends SubsystemBase {
    private final int LEDS_PER_ANIMATION = 30;
    private final CANdle m_candle = new CANdle(Constants.CANdleID, "rio");
    private XboxController joystick;
    private int m_candleChannel = 0;
    private boolean m_clearAllAnims = false;
    private boolean m_last5V = false;
    private boolean m_animDirection = false;
    private boolean m_setAnim = false;

    private Animation m_toAnimate = null;

    public enum AnimationTypes {
        ColorFlow,
        Fire,
        Larson,
        Rainbow,
        RgbFade,
        SingleFade,
        Strobe,
        Twinkle,
        TwinkleOff,
        SetAll,
        Empty
    }
    private AnimationTypes m_currentAnimation;

    public CANdleSystem(XboxController joy) {
        this.joystick = joy;
        changeAnimation(AnimationTypes.SetAll);
        CANdleConfiguration configAll = new CANdleConfiguration();
        configAll.statusLedOffWhenActive = true;
        configAll.disableWhenLOS = false;
        configAll.stripType = LEDStripType.GRB;
        configAll.brightnessScalar = 0.1;
        configAll.vBatOutputMode = VBatOutputMode.Modulated;
        m_candle.configAllSettings(configAll, 100);
    }

    public void toggle5VOverride() {
        System.out.println("State is: " + m_last5V);
        m_candle.configV5Enabled(m_last5V);
        m_last5V = !m_last5V;
    }
    public void toggleAnimDirection() {
        m_animDirection = !m_animDirection;
    }
    public int getMaximumAnimationCount() {
        return m_candle.getMaxSimultaneousAnimationCount();
    }

    public void incrementAnimation() {
        switch(m_currentAnimation) {
            case ColorFlow: changeAnimation(AnimationTypes.Fire); break;
            case Fire: changeAnimation(AnimationTypes.Larson); break;
            case Larson: changeAnimation(AnimationTypes.Rainbow); break;
            case Rainbow: changeAnimation(AnimationTypes.RgbFade); break;
            case RgbFade: changeAnimation(AnimationTypes.SingleFade); break;
            case SingleFade: changeAnimation(AnimationTypes.Strobe); break;
            case Strobe: changeAnimation(AnimationTypes.Twinkle); break;
            case Twinkle: changeAnimation(AnimationTypes.TwinkleOff); break;
            case TwinkleOff: changeAnimation(AnimationTypes.Empty); break;
            case Empty: changeAnimation(AnimationTypes.ColorFlow); break;
            case SetAll: changeAnimation(AnimationTypes.ColorFlow); break;
        }
    }
    public void decrementAnimation() {
        switch(m_currentAnimation) {
            case ColorFlow: changeAnimation(AnimationTypes.Empty); break;
            case Fire: changeAnimation(AnimationTypes.ColorFlow); break;
            case Larson: changeAnimation(AnimationTypes.Fire); break;
            case Rainbow: changeAnimation(AnimationTypes.Larson); break;
            case RgbFade: changeAnimation(AnimationTypes.Rainbow); break;
            case SingleFade: changeAnimation(AnimationTypes.RgbFade); break;
            case Strobe: changeAnimation(AnimationTypes.SingleFade); break;
            case Twinkle: changeAnimation(AnimationTypes.Strobe); break;
            case TwinkleOff: changeAnimation(AnimationTypes.Twinkle); break;
            case Empty: changeAnimation(AnimationTypes.TwinkleOff); break;
            case SetAll: changeAnimation(AnimationTypes.ColorFlow); break;
        }
    }
    public void setColors() {
        changeAnimation(AnimationTypes.SetAll);
    }

    /* Wrappers so we can access the CANdle from the subsystem */
    public double getVbat() { return m_candle.getBusVoltage(); }
    public double get5V() { return m_candle.get5VRailVoltage(); }
    public double getCurrent() { return m_candle.getCurrent(); }
    public double getTemperature() { return m_candle.getTemperature(); }
    public void configBrightness(double percent) { m_candle.configBrightnessScalar(percent, 0); }
    public void configLos(boolean disableWhenLos) { m_candle.configLOSBehavior(disableWhenLos, 0); }
    public void configLedType(LEDStripType type) { m_candle.configLEDType(type, 0); }
    public void configStatusLedBehavior(boolean offWhenActive) { m_candle.configStatusLedState(offWhenActive, 0); }

    public void changeAnimation(AnimationTypes toChange) {
        m_currentAnimation = toChange;
        
        switch(toChange)
        {
            default:
            case ColorFlow:
                m_candleChannel = 0;
                m_toAnimate = new ColorFlowAnimation(128, 20, 70, 0, 0.7, LEDS_PER_ANIMATION, Direction.Forward, m_candleChannel * LEDS_PER_ANIMATION + 8);
                break;
            case Fire:
                m_candleChannel = 1;
                m_toAnimate = new FireAnimation(0.5, 0.7, LEDS_PER_ANIMATION, 0.8, 0.5, m_animDirection, m_candleChannel * LEDS_PER_ANIMATION + 8);
                break;
            case Larson:
                m_candleChannel = 2;
                m_toAnimate = new LarsonAnimation(0, 255, 46, 0, 0.1, LEDS_PER_ANIMATION, BounceMode.Front, 3, m_candleChannel * LEDS_PER_ANIMATION + 8);
                break;
            case Rainbow:
                m_candleChannel = 3;
                m_toAnimate = new RainbowAnimation(1, 0.7, LEDS_PER_ANIMATION, m_animDirection, m_candleChannel * LEDS_PER_ANIMATION + 8);
                break;
            case RgbFade:
                m_candleChannel = 4;
                m_toAnimate = new RgbFadeAnimation(0.7, 0.4, LEDS_PER_ANIMATION, m_candleChannel * LEDS_PER_ANIMATION + 8);
                break;
            case SingleFade:
                m_candleChannel = 5;
                m_toAnimate = new SingleFadeAnimation(50, 2, 200, 0, 0.5, LEDS_PER_ANIMATION, m_candleChannel * LEDS_PER_ANIMATION + 8);
                break;
            case Strobe:
                m_candleChannel = 6;
                m_toAnimate = new StrobeAnimation(240, 10, 180, 0, 0.01, LEDS_PER_ANIMATION, m_candleChannel * LEDS_PER_ANIMATION + 8);
                break;
            case Twinkle:
                m_candleChannel = 7;
                m_toAnimate = new TwinkleAnimation(30, 70, 60, 0, 0.4, LEDS_PER_ANIMATION, TwinklePercent.Percent42, m_candleChannel * LEDS_PER_ANIMATION + 8);
                break;
            case TwinkleOff:
                m_candleChannel = 8;
                m_toAnimate = new TwinkleOffAnimation(70, 90, 175, 0, 0.2, LEDS_PER_ANIMATION, TwinkleOffPercent.Percent76, m_candleChannel * LEDS_PER_ANIMATION + 8);
                break;
            case Empty:
                m_candleChannel = 9;
                m_toAnimate = new RainbowAnimation(1, 0.7, LEDS_PER_ANIMATION, m_animDirection, m_candleChannel * LEDS_PER_ANIMATION + 8);
                break;

            case SetAll:
                m_toAnimate = null;
                break;
        }
        System.out.println("Changed to " + m_currentAnimation.toString());
    }

    public void clearAllAnims() {m_clearAllAnims = true;}

    @Override
    public void periodic() {
        // This method will be called once per scheduler run
        if(m_toAnimate == null) {
            if(!m_setAnim) {
                /* Only setLEDs once, because every set will transmit a frame */
                m_candle.setLEDs(255, 255, 255, 0, 0, 1);
                m_candle.setLEDs(255, 255, 0, 0, 1, 1);
                m_candle.setLEDs(255, 0, 255, 0, 2, 1);
                m_candle.setLEDs(255, 0, 0, 0, 3, 1);
                m_candle.setLEDs(0, 255, 255, 0, 4, 1);
                m_candle.setLEDs(0, 255, 0, 0, 5, 1);
                m_candle.setLEDs(0, 0, 0, 0, 6, 1);
                m_candle.setLEDs(0, 0, 255, 0, 7, 1);
                m_setAnim = true;
            }
        } else {
            m_toAnimate.setSpeed((joystick.getRightY() + 1.0) / 2.0);
            m_candle.animate(m_toAnimate, m_candleChannel);
            m_setAnim = false;
        }
        m_candle.modulateVBatOutput(joystick.getRightY());

        if(m_clearAllAnims) {
            m_clearAllAnims = false;
            for(int i = 0; i < 10; ++i) {
                m_candle.clearAnimation(i);
            }
        }
    }

    @Override
    public void simulationPeriodic() {
        // This method will be called once per scheduler run during simulation
    }
}
